### PR TITLE
[feat]: add COSMOS 2.5 DiT implementation

### PR DESCRIPTION
**Pull request by KyleShao1016**
Implement NVIDIA's COSMOS 2.5 Diffusion Transformer (DiT) architecture in FastVideo
with verified numerical accuracy against the official cosmos-predict2.5 reference.

Key changes:

1. Model Architecture (fastvideo/models/dits/cosmos2_5.py):
   - Implement Cosmos25Transformer3DModel with 28 transformer blocks
   - Add AdaLN-LoRA conditioning with proper modulation parameter computation
   - Implement 3D RoPE with NTK-aware extrapolation for spatiotemporal encoding
   - Add QK normalization using RMSNorm for improved training stability
   - Support optional cross-attention projection for high-dimensional embeddings (Qwen 7B)
   - Add learnable positional embeddings as optional feature
   - Implement proper patch embedding/unpatchification for video processing

2. Configuration (fastvideo/configs/models/dits/cosmos2_5.py):
   - Add Cosmos25ArchConfig with complete parameter mapping for checkpoint loading
   - Map official checkpoint structure (net.blocks.*) to FastVideo structure (transformer_blocks.*)
   - Support AdaLN modulation layers at block level (adaln_modulation_self_attn/cross_attn/mlp)
   - Configure cross-attention projection for 100,352-dim Qwen embeddings
   - Add LoRA parameter mappings for fine-tuning support

3. Tests (fastvideo/tests/transformers/test_cosmos2_5.py):
   - Add comprehensive parity test against official MinimalV1LVGDiT reference
   - Test both single-frame (image) and multi-frame (video) generation
   - Verify checkpoint loading and weight consistency
   - Achieve 0.173% relative difference in bfloat16 (excellent parity)
   - Test video2world conditioning with condition masks

4. Registry (fastvideo/configs/models/dits/__init__.py):
   - Register Cosmos25VideoConfig for model instantiation

Test Results:
- Max difference: 0.046875
- Mean difference: 0.001869
- Relative difference: 0.173% (excellent parity in bfloat16)

### CODE REVIEW

File: fastvideo/tests/transformers/test_cosmos2_5.py [hunk 1]
```diff
@@ -0,0 +1,621 @@
...
+        logger.info(f"Max difference: {max_diff.item():.6f}")
+        logger.info(f"Mean difference: {mean_diff.item():.6f}")
+        logger.info(f"Relative difference: {relative_diff.item():.6f}")
+        
+        # Note: Both models now use the same crossattn_projection (100352 → 1024) with
+        # 100K-dim input embeddings, so they should match much more closely.
+        
+        # Allow for some numerical differences due to implementation details
+        assert max_diff < 1e-1, f"Maximum difference too large: {max_diff.item()}"
+        assert mean_diff < 1e-2, f"Mean difference too large: {mean_diff.item()}"
```
Comment (Edenzzzz) [inline fastvideo/tests/transformers/test_cosmos2_5.py:444]: assert a lower error than this?

File: fastvideo/models/dits/cosmos2_5.py [hunk 2]
```diff
@@ -0,0 +1,940 @@
...
+        key = self.norm_k(key)
+
+        # Apply RoPE if provided (query/key are now in (B, H, S, D_h) format)
+        if rope_emb is not None:
+            cos, sin = rope_emb
+            query = apply_rotary_emb(query, (cos, sin), use_real=True, use_real_unbind_dim=-2)
+            key = apply_rotary_emb(key, (cos, sin), use_real=True, use_real_unbind_dim=-2)
+
+        # Attention computation
+        attn_output = torch.nn.functional.scaled_dot_product_attention(
```
Comment (SolitaryThinker) [inline fastvideo/models/dits/cosmos2_5.py]: can you replace this with fastvideo's distributedAttention? refer to wan or cosmos2. By default both torch sdpa and fa should be supported

File: fastvideo/models/dits/cosmos2_5.py [hunk 3]
```diff
@@ -0,0 +1,940 @@
...
+        query = self.norm_q(query)
+        key = self.norm_k(key)
+
+        # Reshape for attention: (B, S, H, D_h) -> (B, H, S, D_h)
+        query = query.transpose(1, 2)
+        key = key.transpose(1, 2)
+        value = value.transpose(1, 2)
+
+        # Attention computation
+        attn_output = torch.nn.functional.scaled_dot_product_attention(
```
Comment (SolitaryThinker) [inline fastvideo/models/dits/cosmos2_5.py]: replace this with our LocalAttention layer.

**Comment by SolitaryThinker** [comment]
also please run pre-commit and fix lint

File: fastvideo/models/dits/cosmos2_5.py [hunk 4]
```diff
@@ -0,0 +1,940 @@
...
+
+        batch_size, num_channels, num_frames, height, width = hidden_states.shape
+
+        # 1. Concatenate condition mask if provided
+        if condition_mask is not None:
+            hidden_states = torch.cat([hidden_states, condition_mask], dim=1)
+
+        # 2. Concatenate padding mask if needed
+        if self.concat_padding_mask and padding_mask is not None:
+            from torchvision import transforms
```
[the remark this answers is not in the document]
Comment (KyleShao1016) [reply fastvideo/models/dits/cosmos2_5.py]: Moved the import code to the top section.

File: fastvideo/models/dits/cosmos2_5.py [hunk 5]
```diff
@@ -0,0 +1,940 @@
...
+        hidden_size: int,
+        max_size: tuple[int, int, int] = (128, 240, 240),
+        patch_size: tuple[int, int, int] = (1, 2, 2),
+        base_fps: int = 24,
+        rope_scale: tuple[float, float, float] = (1.0, 1.0, 1.0),
+        enable_fps_modulation: bool = True,
+    ) -> None:
+        super().__init__()
+
+        self.max_size = [size // patch for size, patch in zip(max_size, patch_size, strict=False)]
```
[the remark this answers is not in the document]
Comment (KyleShao1016) [reply fastvideo/models/dits/cosmos2_5.py]: Updated!

File: fastvideo/models/dits/cosmos2_5.py [hunk 6]
```diff
@@ -0,0 +1,940 @@
...
+    def __init__(
+        self,
+        hidden_size: int,
+        max_size: tuple[int, int, int],
+        patch_size: tuple[int, int, int],
+        eps: float = 1e-6,
+    ) -> None:
+        super().__init__()
+
+        self.max_size = [size // patch for size, patch in zip(max_size, patch_size, strict=False)]
```
[the remark this answers is not in the document]
Comment (KyleShao1016) [reply fastvideo/models/dits/cosmos2_5.py]: Updated!

File: fastvideo/tests/transformers/test_cosmos2_5.py [hunk 7]
```diff
@@ -0,0 +1,621 @@
...
+COSMOS_PREDICT2_5_PATH = os.path.join(TEST_DIR, '..', '..', '..', '..', 'cosmos-predict2.5')
+COSMOS_PREDICT2_5_PATH = os.path.normpath(COSMOS_PREDICT2_5_PATH)
+if os.path.exists(COSMOS_PREDICT2_5_PATH) and COSMOS_PREDICT2_5_PATH not in sys.path:
+    sys.path.insert(0, COSMOS_PREDICT2_5_PATH)
+
+from fastvideo.configs.pipelines import PipelineConfig
+from fastvideo.forward_context import set_forward_context
+from fastvideo.fastvideo_args import FastVideoArgs
+from fastvideo.logger import init_logger
+from fastvideo.models.loader.component_loader import TransformerLoader
```
[the remark this answers is not in the document]
Comment (KyleShao1016) [reply fastvideo/tests/transformers/test_cosmos2_5.py]: Removed!

File: fastvideo/tests/transformers/test_cosmos2_5.py (hunk 1 above)
Comment (KyleShao1016) [reply fastvideo/tests/transformers/test_cosmos2_5.py:444]: Changed the `dtype` to float32 and reduced the threshold values.

File: fastvideo/models/dits/cosmos2_5.py (hunk 2 above)
Comment (KyleShao1016) [reply fastvideo/models/dits/cosmos2_5.py]: Updated!

File: fastvideo/models/dits/cosmos2_5.py (hunk 3 above)
Comment (KyleShao1016) [reply fastvideo/models/dits/cosmos2_5.py]: Updated!

File: fastvideo/models/dits/cosmos2_5.py [hunk 8]
```diff
@@ -0,0 +1,1002 @@
...
+            supported_attention_backends = (AttentionBackendEnum.FLASH_ATTN, AttentionBackendEnum.TORCH_SDPA)
+        
+        # Check if distributed environment is initialized
+        try:
+            from fastvideo.distributed.parallel_state import model_parallel_is_initialized
+            use_distributed = torch.distributed.is_initialized() and model_parallel_is_initialized()
+        except:
+            use_distributed = False
+        
+        if use_distributed:
```
Comment (SolitaryThinker) [inline fastvideo/models/dits/cosmos2_5.py]: we shouldn't need to check this, just always using distributedAttention is fine

Comment (KyleShao1016) [reply fastvideo/models/dits/cosmos2_5.py]: Updated!
I used DistributedAttention for self attention and LocalAttention for cross attention.
